README.md
# IVAR - Lua in Redis debugger and monitoring tool

![debugger demo](https://github.com/tsykhovskyi/ivar/blob/master/samples/debugger_demo.png)

## Installation

Install package globally or as dev dependency

```
npm i -g @tsykhovskyi/ivar
```

## Usage

Start debugger in proxy mode. Set up forward traffic to redis port (default: 6379).

```
ivar proxy --tunnel 6380:6379 --sync-mode
```

If debugger is installed as a project dependency next npm script can be defined

```
npx ivar proxy -t 6380:6379 --sync-mode
```

Then run from project root

```bash
npm run redis:debugger
```

Debugger opens additional redis port(`6380`) and will forward all traffic from it to the main
redis server. By default, debugger is enabled so it will run debug process on each
`EVAL` command.

Let's try with simple example

```bash
# Login with redis-cli REPL to debugger proxy port
redis-cli -p 6380

# Run script
EVAL "local msg = ARGV[1] .. ARGV[2];\n return msg" 0 "Hello" "world"
```

Console is halted now and debug window can be opened at http://127.0.0.1:29999. After script
execution is finished console will return its result.

## Features

This application is a HTTP server that provide handy UI for Lua in Redis scripts debugging.
Two modes are available for debugging:

- _proxy mode_: forwards commands to redis server and run script debugger on demand
- _command mode_: Lua script file is executed with arguments

In proxy mode debugger you can use debugger ports as redis for your application. If Redis is
set up in Cluster mode, then for better experience and debug coverage you should forward ports
for all nodes.

Debugger supports both Redis `forked`(default) and `sync` redis debugging mode. It can be changed
in UI and predefined with `--sync-mode` option. https://redis.io/docs/manual/programmability/lua-debugging/#synchronous-mode

### Debugger UI features

- Step (F8)
- Continue (F9)
- Abort/Restart execution
- Breakpoints
- Scope/Watch variables
- Multiple debug session simultaneously

## Debugger CLI help

### Debugger server

```
ivar [command]

Commands:
  ivar eval <file> <keys-and-args...>  send an EVAL command using the Lua script
                                               at <file>.
  ivar proxy                           run tcp proxy and debug traffic

Options:
      --version    Show version number                                           [boolean]
  -p, --port       debugger server port                          [number] [default: 29999]
      --sync-mode  run debugger in sync mode                    [boolean] [default: false]
      --help       Show help                                                     [boolean]

Examples:
  ivar --port 80    Start debugger UI on port 80
  ivar --sync-mode  Start debugger in sync mode
  ivar eval --help  Show eval command help
  ivar tcp --help   Show tcp command help

```

### Proxy mode

Allow to forward Redis port traffic and intercept required Lua scripts

```
ivar proxy

run tcp proxy and debug traffic

Options:
      --version    Show version number                                           [boolean]
  -p, --port       debugger server port                          [number] [default: 29999]
      --sync-mode  run debugger in sync mode                    [boolean] [default: false]
      --help       Show help                                                     [boolean]
  -t, --tunnel     redis port paths map                                 [array] [required]
  -f, --filter     debug scripts that match filter                                 [array]
      --disable    disable debugger on start                    [boolean] [default: false]

Examples:
  ivar proxy --tunnel 6380:6379          Open proxy port with traffic forward and
  --filter <keyword>                             start debugger if <keyword> occurs
  ivar proxy --t 6380:6379 --sync-mode   Start debug sessions in LDB sync mode
  ivar proxy --tunnel 6380:6379          Disable traffic interception by default
  --disable

```

### Command mode

Allow to debug single Lua script file

```
ivar eval <file> <keys-and-args...>

send an EVAL command using the Lua script at <file>.

Positionals:
  file           lua file path                                         [string] [required]
  keys-and-args  comma separated keys and args            [array] [required] [default: []]

Options:
      --version     Show version number                                          [boolean]
  -p, --port        debugger server port                         [number] [default: 29999]
      --sync-mode   run debugger in sync mode                   [boolean] [default: false]
      --help        Show help                                                    [boolean]
  -P, --redis-port  redis server port                  [number] [required] [default: 6379]

Examples:
  ivar eval ./script.lua -P 6379      Debug local script file with redis on port
                                              6379
  ivar eval ./script.lua key1 , arg1  Debug script with key and argument

```

## Examples

### Proxy mode

Launch redis proxy with disabled debugging. It can be activated on UI.

```bash
ivar proxy --tunnel 6380:6379 --disabled
```

Launch multiple redis proxies in sync mode. Sync/forked modes can be changed in UI.
In cluster mode it is recommended to forward all redis node ports.

```bash
ivar proxy --tunnel 63790:6379 --tunnel 63800:6380 --tunnel 63810:6381 --sync-mode
```

#### Command mode

Execute `~/scripts/test.lua` with a non-default redis port

```bash
ivar eval ~/scripts/test.lua --redis-port 30000
```

Execute `~/scripts/test.lua` file in sync mode with keys (`key1`, `key2`)
and arguments (`veni` `vidi` `vici`)

```bash
ivar eval ~/scripts/test.lua key1 key2 , veni vidi vici --sync-mode
```


CHANGELOG.md
# ChangeLog

## [0.0.14] - 2022-02-27

### Fixed

- Save lua script eval even in debug disabled mode

## [0.0.13] - 2022-01-24

### Fixed

- UI debugger counter was not updated


## [0.0.12] - 2022-01-18

### Changed

- UI request responsiveness fix
- UI active tunnels & github button
- auto-save script in memory on `eval` command

### Fixed

- handle multichunk requests by proxy

## [0.0.11] - 2022-01-17

### Fixed

- handle multichunk requests by proxy

## [0.0.10] - 2022-01-11

### Changed

- three available modes for traffic requests: raw, pretty, grouped
- prevent fonts loading via network (only local files)
- traffic size per request

## [0.0.9] - 2022-01-10

### Changed

- Traffic proxy ports monitoring for both redis and debugger
- UI/UX improvements for traffic preview
- Removed raw preview for redis requests(UI)

### Fixed

- bugfix: interceptor string/bulk-string did not trigger

## [0.0.8] - 2022-01-09

### Changed

- request structure both RESP and plain style support
- `redis-cli --pipe` support
- pending request preview (UI)
- traffic "clear" button (UI)
- RESP protocol tests coverage

### Fixed

- binary safe data transfer
- debugger result fixed view
- UI bugfixes

## [0.0.7] - 2022-12-30

### Changed

- preview(redis-cli style) for request/response data structure

### Fixed

- show UI debug toolbar only on running session

## [0.0.6] - 2022-12-28

### Changed

- save `eval load` script in debugger memory instead of reject `evalsha` requests from client
- added debugger session finished result UI
- support for non-RESP requests
- proxy traffic monitoring, request/response preview 

## [0.0.5] - 2022-12-19

### Changed

- hide traffic in logs and mention only byte size and request payload

### Fixed

- fix debugger "restart" command hang up 

## [0.0.4] - 2022-12-17

### Changed

- Prevent skipping of nodes debugger ports by ports substitution in redis response 

### Fixed

- `Eval` mode throw exception on script fail
- Fixed redis message parsing in multiple TCP chunks

## [0.0.3] - 2022-12-12

### Changed

- Redis Cluster support
- Redis response parser logic abstracted from TCP chunks

## [0.0.2] - 2022-12-10

### Changed

- Enhance code style with _eslint_ and _prettier_

### Fixed

- Fix debugger is not responding when invalid script executed


src/ivar.ts
#!/usr/bin/env node

import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { evalCommand } from './comands/eval.command';
import { proxyCommand } from './comands/proxy.command';
import { serverCommand } from './comands/server.command';

(async () => {
  yargs(hideBin(process.argv))
    .option('port', {
      alias: 'p',
      default: 29999, // todo check 0 port
      type: 'number',
      description: 'debugger server port',
    })
    .option('sync-mode', {
      default: false,
      type: 'boolean',
      description: 'run debugger in sync mode',
    })
    .middleware((args) => {
      serverCommand.handle(args);
    })
    .command(
      'eval <file> <keys-and-args...>',
      'send an EVAL command using the Lua script at <file>.',
      (yargs) =>
        yargs
          .positional('file', {
            description: 'lua file path',
            type: 'string',
            demandOption: true,
          })
          .positional('keys-and-args', {
            description: 'comma separated keys and args',
            type: 'string',
            array: true,
          })
          .option('redis-port', {
            alias: 'P',
            description: 'redis server port',
            type: 'number',
            default: 6379,
            demandOption: true,
          })
          .example([
            [
              '$0 eval ./script.lua -P 6379',
              'Debug local script file with redis on port 6379',
            ],
            [
              '$0 eval ./script.lua key1 , arg1',
              'Debug script with key and argument',
            ],
          ]),
      async (args) => {
        const result = await evalCommand.handle(args);
        console.log('Script finished with result\n');
        console.log(JSON.stringify(result));
        process.exit(0);
      }
    )
    .command(
      'proxy',
      'run tcp proxy and debug traffic',
      (yargs) =>
        yargs
          .option('tunnel', {
            alias: 't',
            type: 'string',
            array: true,
            describe: 'redis port paths map',
          })
          .option('filter', {
            alias: 'f',
            type: 'string',
            array: true,
            describe: 'debug scripts that match filter',
          })
          .option('disable', {
            default: false,
            type: 'boolean',
            description: 'disable debugger on start',
          })
          .demandOption(['tunnel'], 'At least one tunnel should be defined')
          .example([
            [
              '$0 proxy --tunnel 6380:6379 --filter <keyword>',
              'Open proxy port with traffic forward and start debugger if <keyword> occurs',
            ],
            [
              '$0 proxy --t 6380:6379 --sync-mode',
              'Start debug sessions in LDB sync mode',
            ],
            [
              '$0 proxy --tunnel 6380:6379 --disable',
              'Disable traffic interception by default',
            ],
          ]),
      (args) => proxyCommand.handle(args)
    )
    .recommendCommands()
    .help()
    .example([
      ['$0 --port 80', 'Start debugger UI on port 80'],
      ['$0 --sync-mode', 'Start debugger in sync mode'],
      ['$0 eval --help', 'Show eval command help'],
      ['$0 tcp --help', 'Show tcp command help'],
    ])
    .wrap(90)
    .recommendCommands()
    .showHelpOnFail(true)
    .scriptName('ivar').argv;

  process.on('uncaughtException', function (err) {
    console.error('UNCAUGHT EXCEPTION:', err);
  });
})();
